05.ls/ls.rb

```diff
@@ -0,0 +1,94 @@
+#!/usr/bin/env ruby
+# frozen_string_literal: true
+
+require 'etc'
+require 'optparse'
+require 'date'
+
+def main
+  reverse_flag = list_flag = 0
+  OptionParser.new do |opt|
+    items, hidden_items = recieve_items
+    opt.on('-a') { items += hidden_items }
+    opt.on('-r') { reverse_flag = 1 }
+    opt.on('-l') { list_flag = 1 }
+    opt.parse!(ARGV)
+    sorted_items = reverse_flag == 1 ? items.sort.reverse : items.sort
+    list_flag == 1 ? list_process(sorted_items) : no_list_process(sorted_items)
+  end
+end
+
+def recieve_items
+  items = []
+  hidden_items = []
+  Dir.foreach('.') do |item|
+    item[0] == '.' ? hidden_items << item : items << item
+  end
+  [items, hidden_items]
+end
+
+def list_process(items)
+  list_stats = []
+  all_block = 0
+  items.each do |item|
+    item_state = File.stat(item)
+    all_block += item_state.blocks
+    list_stats << "#{define_list_variable(item_state)} #{item}"
+  end
+  print_list(all_block, list_stats)
+end
+
+def print_list(block, items)
+  puts "total #{block}"
+  items.each { |item| puts item }
+end
+
+def define_list_variable(item_state)
+  file_type = item_state.ftype == 'file' ? '-' : 'd'
+  file_permission = change_permisson(item_state.mode.to_s(8))
+  link_counts = item_state.nlink
+  owner_name = Etc.getpwuid(item_state.uid).name
+  group_name = Etc.getgrgid(item_state.gid).name
+  item_size = item_state.size
+  time = timestamp(item_state)
+  "#{file_type}#{file_permission}#{format(link_counts, 3)}#{format(owner_name, 6)}#{format(group_name, 7)}#{format(item_size, 6)}#{time}"
+end
+
+def timestamp(item_state)
+  month = item_state.mtime.month
+  day = item_state.mtime.day
+  hour = item_state.mtime.hour
+  minute = item_state.mtime.min
+  "#{format(month, 3)}#{format(day, 3)}#{format("#{hour}:#{minute}", 6)}"
+end
+
+def format(text, num)
+  text.to_s.rjust(num)
+end
+
+def change_permisson(text)
+  permission = ''
+  correspondence = { '0' => '---', '1' => '--x', '2' => '-w-', '3' => '-wx',
+                     '4' => 'r--', '5' => 'r-x', '6' => 'rw-', '7' => 'rwx' }
```

Comment:
こんな感じで書いた方が視認性良さそうです。

```suggestion
  correspondence = {
    '0' => '---',
    '1' => '--x',
    '2' => '-w-',
    '3' => '-wx',
    '4' => 'r--',
    '5' => 'r-x',
    '6' => 'rw-',
    '7' => 'rwx'
  }
```

また、こういうキー/バリューのペアでマッピングを表すようなデータ構造はmap, table, dictionaryなどと呼ばれます。hashとも言いますね（データ型の名前的にはhashですね）。

こういう変数に変数名をつけるときには、こういった単語を利用してもいいかもしれません。

Reply:
修正しました。行を削ることばかりに気を取られ、視認性への配慮ができていませんでした。